drivers/src/pmu.rs

```diff
@@ -0,0 +1,134 @@
+            // but the presence of a firmware counter indicates HW counters are exhausted.
+            if info.is_firmware_counter() {
+                return Ok(ControlFlow::Break(()));
+            } else if !Self::is_valid_hardware_counter(info) {
+                return Err(SbiError::NotSupported);
+            }
+            self.hw_counters_info.push(Some(info));
+            self.valid_hw_counters_mask |= 1 << counter_index;
+        } else {
+            // Hardware counters can be sparse, and may not be implemented.
+            self.hw_counters_info.push(None);
+        }
+        Ok(ControlFlow::Continue(()))
+    }
+
+    /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
+    pub fn init() -> SbiResult<()> {
+        let mut pmu_info = PmuInfo::new();
+        let num_counters = pmu::get_num_counters()?;
+        for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
+            if matches!(pmu_info.init_counter(i)?, ControlFlow::Break(())) {
+                break;
+            }
+        }
+        if !pmu_info.hw_counters_info.is_empty() {
+            PMU_INFO.call_once(|| pmu_info);
+            Ok(())
+        } else {
+            Err(SbiError::NotSupported)
+        }
+    }
+
+    /// Returns a reference to the global PmuInfo structure.
+    pub fn get() -> SbiResult<&'static PmuInfo> {
+        PMU_INFO.get().ok_or(SbiError::NotSupported)
+    }
+
+    /// Returns the counter index for the specified CSR
+    pub fn csr_to_counter_index(&self, csr: u64) -> SbiResult<u64> {
+        let counter_index = self
+            .hw_counters_info
+            .iter()
+            .position(|info| info.filter(|i| i.get_csr() == csr).is_some())
+            .ok_or(SbiError::InvalidParam)? as u64;
+        Ok(counter_index)
+    }
+
+    /// Returns the number of hardware counters supported by the platform
+    pub fn get_num_counters(&self) -> u64 {
+        self.hw_counters_info.len() as u64
+    }
+
+    /// Returns cached information for the hardware counter specified by counter_index.
+    pub fn get_counter_info(&self, counter_index: u64) -> SbiResult<PmuCounterInfo> {
+        let info = self
+            .hw_counters_info
+            .get(counter_index as usize)
+            .ok_or(SbiError::InvalidParam)?;
+        if let Some(info) = info {
+            Ok(*info)
+        } else {
+            Err(SbiError::InvalidParam)
+        }
+    }
+
+    /// Returns the CSR for the specified counter index
+    pub fn counter_index_to_csr(&self, counter_index: u64) -> SbiResult<u64> {
+        Ok(self.get_counter_info(counter_index)?.get_csr())
+    }
+
+    /// Validates the counter_index and counter_mask bit-mask for counter start and stop operations.
+    /// The counter_mask is relative to counter_index. The return value is a sanitized version
+    /// of the original counter_mask, with set bits representing implemented platform counters.
+    pub fn validate_counter_and_mask(
```

Comment:
Maybe `filter_counter_mask` is a more appropriate name given the new behavior?

Reply:
Done.